contracts/GroupVoteLocker.sol

```diff
@@ -0,0 +1,331 @@
+             * block since the last checkpoint and present blockchain state. 
+             */
+            dTime =
+                ((_blockNumber - checkpoint0.blk) *
+                    (block.timestamp - checkpoint0.ts)) /
+                (block.number - checkpoint0.blk);
+        }
+        // if code doesn't enter any of the above if conditions latest _blockNumber was passed
+        // to the function and dTime is correctly set to 0
+
+        // Now dTime contains info on how far are we beyond point
+        return _votePowerAt(groupState, checkpoint0, checkpoint0.ts + dTime);
+    }
+
+    /**
+     * @dev Calculates voting power at a given time _t
+     * @param groupState group state to calculate voting power from
+     * @param _checkpoint Most recent point before time _t
+     * @param _time Time at which to calculate supply
+     * @return totalSupply at given time
+     */
+    function _votePowerAt(GroupVotePowerState storage groupState, Checkpoint memory _checkpoint, uint256 _time)
+        internal
+        view
+        returns (uint256)
+    {
+        Checkpoint memory lastCheckpoint = _checkpoint;
+
+        // Floor the timestamp to weekly interval
+        uint256 iterativeTime = _floorToWeek(lastCheckpoint.ts);
+        // Iterate through all weeks between _checkpoint & _time to account for slope changes
+        for (uint256 i = 0; i < 255; i++) {
+            iterativeTime = iterativeTime + WEEK;
+            int128 dSlope = 0;
+            // If week end is after timestamp, then truncate & leave dSlope to 0
+            if (iterativeTime > _time) {
+                iterativeTime = _time;
+            }
+            // else get most recent slope change
+            else {
+                dSlope = groupState.slopeChanges[iterativeTime];
+            }
+
+            lastCheckpoint.bias =
+                lastCheckpoint.bias -
+                (lastCheckpoint.slope *
+                    SafeCast.toInt128(
+                        int256(iterativeTime - lastCheckpoint.ts)
+                    ));
+
+            if (iterativeTime == _time) {
+                break;
+            }
+
+            lastCheckpoint.slope = lastCheckpoint.slope + dSlope;
+            lastCheckpoint.ts = iterativeTime;
+        }
+
+        return SafeCast.toUint256(max(lastCheckpoint.bias, 0));
+    }
+
+    /**
+     * @dev Binary search (bisection) to find epoch closest to block.
+     * @param _block Find the most recent point history before this block
+     * @param _maxEpoch Maximum epoch
+     * @return uint256 The most recent epoch before the block
+     */
+    function _findEpoch(
+        Checkpoint[] memory _checkpoints,
+        uint256 _block,
+        uint256 _maxEpoch
+    ) internal view returns (uint256) {
+        uint256 minEpoch = 0;
+        uint256 maxEpoch = _maxEpoch;
+        for (uint256 i = 0; i < 128; i++) {
+            if (minEpoch >= maxEpoch) break;
```

Comment:
As far as I can tell, this searches "from the bottom up". The most expensive gas search would be for search for the most recent checkpoint. 

If this were used for searching future checkpoints, this would be efficient, but it looks like when it's used, it required that it is searching older blocks.

Sill correct, may or may not make much difference, just something I noticed.

Reply:
I think this uses bisection for searching